StRoot/StFcsDbMaker/StFcsDb.cxx

```diff
@@ -1979,3 +1979,125 @@ void StFcsDb::readGainCorrFromText(){
   }
   fclose(F);
 }
+
+//g2t track info                                                                                                   
+unsigned int StFcsDb::backTraceG2tTrack(unsigned int id, g2t_track_st* g2ttrk){
+  int i = id - 1;
+  while(g2ttrk[i].next_parent_p !=0){
+    if(mDebug>3)
+      LOG_INFO<<Form("  BackTrace from=%3d id=%3d Epid=%4d Gpid=%3d Vtx=%3d Parent=%3d E=%6.2f",
+                     id,g2ttrk[i].id,g2ttrk[i].eg_pid,g2ttrk[i].ge_pid,g2ttrk[i].start_vertex_p,
+		     g2ttrk[i].next_parent_p,g2ttrk[i].e)<<endm;
+    i = g2ttrk[i].next_parent_p - 1;
+  }
+  if(mDebug>3) LOG_INFO<<Form("  BackTrace from=%3d id=%3d Epid=%4d Gpid=%3d Vtx=%3d Parent=%3d E=%6.2f Primary!!",
+			      id,g2ttrk[i].id,g2ttrk[i].eg_pid,g2ttrk[i].ge_pid,g2ttrk[i].start_vertex_p,
+			      g2ttrk[i].next_parent_p,g2ttrk[i].e)<<endm;
+  return i + 1;
+}
+    
+const g2t_track_st* StFcsDb::getParentG2tTrack(StFcsHit* h, g2t_track_st* g2ttrk, float& fraction, int& ntrk, unsigned int order){
+  const vector<pair<unsigned int, float>>& gt = h->getGeantTracks();  
+  ntrk=gt.size();
+  float detot=0;
+  if(order > gt.size()) {fraction=0; return 0;}
+  for(int itrk=0; itrk<ntrk; itrk++) detot += gt[itrk].second;
+  fraction = gt[order].second / detot;
+  if(mDebug>3){
+    for(unsigned int jtrk=0; jtrk<ntrk; jtrk++){
+      LOG_INFO << Form("Hit's parent G2T Track %3d id=%3d dE=%f",jtrk,gt[jtrk].first,gt[jtrk].second)<<endm;
+    }
+  }
+  return &g2ttrk[gt[order].first-1];
+}
+
+const g2t_track_st* StFcsDb::getPrimaryG2tTrack(StFcsHit* h, g2t_track_st* g2ttrk, float& fraction, int& ntrk,unsigned int order){
+  const vector<pair<unsigned int, float>>& gt = h->getGeantTracks();
+  vector<pair<unsigned int, float>> primary;
+  int n=gt.size();
+  float detot=0;
+  for(int itrk=0; itrk<n; itrk++){
+    float de=gt[itrk].second;
+    unsigned int id=backTraceG2tTrack(gt[itrk].first, g2ttrk);
+    int found=0;
+    unsigned int np=primary.size();
+    for(unsigned int jtrk=0; jtrk<np; jtrk++){
+      if(primary[jtrk].first == id) {primary[jtrk].second += de; found=1; break;}
+    }
+    if(found==0) primary.push_back(make_pair(id,de));
+    detot+=de;
+  }
+  ntrk = primary.size();
+  if(order > ntrk) {fraction=0; return 0;}
+  std::sort(primary.begin(), primary.end(),
+	    [](const pair<unsigned int,float>&a, const pair<unsigned int,float>&b){
+	      return b.second < a.second;
+	    });
+  fraction = primary[order].second / detot;
+  return &g2ttrk[primary[order].first-1];
+}
+
+const g2t_track_st* StFcsDb::getParentG2tTrack(StFcsCluster* c, g2t_track_st* g2ttrk, float& fraction, int& ntrk, unsigned int order){
```

Comment:
I think it would be good to document somewhere what `order` means in these functions. It appears to me that invoking these functions with different values of `order` will be mean repeating all the code as nothing about the parents is stored/cached, so I'd really like to better understand the use cases of `order` to understand if this is the best way to encode these functions.

Thanks,
-Gene

Reply:
- It is in commet in StFcsDb.h file
   - With order=0 returns the highest contributing g2t_track_id
   - With order=1 returns the 2nd highest contributing g2t_track_id
   - Yes, I can make sorted vector available to user....But then user need to delete them.
   - Or cache sorted results somehow, but then delete when & where?
   - Any suggestions will be welcomed